books/kestrel/c/transformation/deftrans.lisp

```diff
@@ -0,0 +1,1927 @@
+         ,(deftrans-defn-enumer              names bodies extra-args extra-args-names)
+         ,(deftrans-defn-enumer-list         names bodies extra-args extra-args-names)
+         ,(deftrans-defn-statassert          names bodies extra-args extra-args-names)
+         :hints (("Goal" :in-theory '(deftrans-measure-theory)))
+         :verify-guards nil
+         ///
+         (verify-guards ,(cdr (assoc-eq 'expr names))))
+       ,(deftrans-defn-initdeclor      names bodies extra-args extra-args-names)
+       ,(deftrans-defn-initdeclor-list names bodies extra-args extra-args-names)
+       ,(deftrans-defn-decl            names bodies extra-args extra-args-names)
+       ,(deftrans-defn-decl-list       names bodies extra-args extra-args-names)
+       ,(deftrans-defn-label           names bodies extra-args extra-args-names)
+       (defines ,name-stmt/blocks
+         ,(deftrans-defn-stmt            names bodies extra-args extra-args-names)
+         ,(deftrans-defn-block-item      names bodies extra-args extra-args-names)
+         ,(deftrans-defn-block-item-list names bodies extra-args extra-args-names)
+         :hints (("Goal" :in-theory '(deftrans-measure-theory)))
+         :verify-guards nil
+         ///
+         (verify-guards ,(cdr (assoc-eq 'stmt names))))
+       ,(deftrans-defn-fundef       names bodies extra-args extra-args-names)
+       ,(deftrans-defn-extdecl      names bodies extra-args extra-args-names)
+       ,(deftrans-defn-extdecl-list names bodies extra-args extra-args-names)
+       ,(deftrans-defn-transunit    names bodies extra-args extra-args-names)
+       (define ,(cdr (assoc-eq 'filepath-transunit-map names))
+         ((map filepath-transunit-mapp)
+          ,@extra-args)
+         :returns (new-map filepath-transunit-mapp
+                           :hyp (filepath-transunit-mapp map))
+         (b* (((when (omap::emptyp map)) nil)
+              ((mv path tunit) (omap::head map))
+              (new-path (deftrans-filepath path ,(symbol-name name)))
+              (new-tunit (,(cdr (assoc-eq 'transunit names)) tunit ,@extra-args-names))
+              (new-map (,(cdr (assoc-eq 'filepath-transunit-map names)) (omap::tail map) ,@extra-args-names)))
+           (omap::update new-path new-tunit new-map))
+         :verify-guards :after-returns)
+       (define ,(cdr (assoc-eq 'transunit-ensemble names))
+         ((tunits transunit-ensemblep)
+          ,@extra-args)
+         :returns (new-tunits transunit-ensemblep)
+         :short "Transform a translation unit ensemble."
+         (b* (((transunit-ensemble tunits) tunits))
+           (transunit-ensemble (,(cdr (assoc-eq 'filepath-transunit-map names)) tunits.unwrap ,@extra-args-names))))))
+  :guard-hints (("Goal" :in-theory (enable atom-listp))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; Separates a list into pairs.
+;; E.g. '(foo bar baz qux) becomes '((foo . bar) (baz . qux)).
+(define take-pairs
```

Comment:
I would move this function early in the file, after the book inclusions and before the deftrans-specific code, since this is really a library extension. So it is easier to find and eventually move to a library.